src/main/java/org/wildfly/extension/elytron/PolicyDefinitions.java

```diff
@@ -0,0 +1,574 @@
+                .setMinSize(1)
+                .build();
+        static final SimpleAttributeDefinition MODULE = ClassLoadingAttributeDefinitions.MODULE;
+        static ObjectTypeAttributeDefinition POLICY = new ObjectTypeAttributeDefinition.Builder(JACC_POLICY, NAME, POLICY_PROVIDER, CONFIGURATION_FACTORY, MODULE).build();
+        static final ObjectListAttributeDefinition POLICIES = new ObjectListAttributeDefinition.Builder(JACC_POLICY, POLICY)
+                .setMinSize(1)
+                .setAllowNull(true)
+                .build();
+    }
+
+    static class CustomPolicyDefinition {
+        static final SimpleAttributeDefinition NAME = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.NAME, ModelType.STRING, false)
+                .setMinSize(1)
+                .build();
+        static final SimpleAttributeDefinition CLASS_NAME = ClassLoadingAttributeDefinitions.CLASS_NAME;
+        static final SimpleAttributeDefinition MODULE = ClassLoadingAttributeDefinitions.MODULE;
+        static ObjectTypeAttributeDefinition POLICY = new ObjectTypeAttributeDefinition.Builder(ElytronDescriptionConstants.CUSTOM_POLICY, NAME, CLASS_NAME, MODULE).build();
+        static final ObjectListAttributeDefinition POLICIES = new ObjectListAttributeDefinition.Builder(ElytronDescriptionConstants.CUSTOM_POLICY, POLICY)
+                .setAllowNull(true)
+                .build();
+    }
+
+    static ResourceDefinition getPolicy() {
+        AttributeDefinition[] attributes = new AttributeDefinition[] {DEFAULT_POLICY, JaccPolicyDefinition.POLICIES, CustomPolicyDefinition.POLICIES};
+        AbstractAddStepHandler add = new BaseAddHandler(POLICY_RUNTIME_CAPABILITY, attributes) {
+            @Override
+            protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+                String defaultPolicy = context.getCurrentAddress().getLastElement().getValue();
+                ServiceName serviceName = POLICY_RUNTIME_CAPABILITY.getCapabilityServiceName(Policy.class);
+                InjectedValue<Supplier<Policy>> policyProviderInjector = new InjectedValue<>();
+                ServiceTarget serviceTarget = context.getServiceTarget();
+                ServiceBuilder<Policy> serviceBuilder = serviceTarget.addService(serviceName, createPolicyService(policyProviderInjector));
+                Supplier<Policy> policySupplier = getPolicyProvider(context, model, defaultPolicy, serviceBuilder);
+
+                policyProviderInjector.setValue(() -> policySupplier);
+
+                serviceBuilder.setInitialMode(Mode.ACTIVE).install();
+
+                if (!context.isBooting()) {
+                    context.reloadRequired();
+                }
+            }
+
+            private Service<Policy> createPolicyService(InjectedValue<Supplier<Policy>> injector) {
+                return new Service<Policy>() {
+                    Policy delegated;
```

Comment:
The variables manipulated by the service lifecycle methods need to be marked as being volatile.